### PR TITLE
Implement a basic PoC of an injectable WebDriver

### LINKED ISSUE #30376: Testsuite PoC - Basic injection of Selenium WebDriver
The test framework should have a basic annotation to inject a Selenium WebDriver into a test. For the first version this can be a hard-coded driver such as Firefox driver.

### PR DESCRIPTION
**Pull request by jonkoops**
Implements a basic injectable WebDriver with a sample test for the welcome page. This only implements the Firefox driver until we have a system in place to allow multiple suppliers of an implementation (see ##30377). This does not yet implement page objects in it's final form, which will be done under #42.

Closes #30376

### CODE REVIEW

File: test-poc/base/src/test/java/org/keycloak/test/base/WelcomePageTest.java [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+import org.keycloak.test.framework.KeycloakIntegrationTest;
+import org.keycloak.test.framework.page.WelcomePage;
+import org.keycloak.test.framework.server.NoAdminUserKeycloakTestServerConfig;
+import org.keycloak.test.framework.webdriver.InjectWebDriver;
+import org.openqa.selenium.WebDriver;
+
+@KeycloakIntegrationTest(config = NoAdminUserKeycloakTestServerConfig.class)
+public class WelcomePageTest {
+
+    @InjectWebDriver
```
Comment (stianst) [inline test-poc/base/src/test/java/org/keycloak/test/base/WelcomePageTest.java:13]: Would we want some standard convention for these annotations? To follow naming conventions elsewhere I'd say this would be `@TestWebDriver`. We could also have a generic annotation for those cases where the annotation doesn't need any configuring, but not sure how common that would be.

File: test-poc/framework/src/main/java/org/keycloak/test/framework/server/NoAdminUserKeycloakTestServerConfig.java [hunk 2]
```diff
@@ -0,0 +1,3 @@
+package org.keycloak.test.framework.server;
+
+public class NoAdminUserKeycloakTestServerConfig implements KeycloakTestServerConfig { }
```
Comment (stianst) [inline test-poc/framework/src/main/java/org/keycloak/test/framework/server/NoAdminUserKeycloakTestServerConfig.java:3]: What's the purpose of this?

File: test-poc/framework/src/main/java/org/keycloak/test/framework/server/DefaultKeycloakTestServerConfig.java [hunk 3]
```diff
@@ -1,5 +1,17 @@
 package org.keycloak.test.framework.server;
 
+import java.util.Optional;
+
 public class DefaultKeycloakTestServerConfig implements KeycloakTestServerConfig {
 
+    @Override
```
Comment (stianst) [inline test-poc/framework/src/main/java/org/keycloak/test/framework/server/DefaultKeycloakTestServerConfig.java:7]: Not sure you'd ever want to configure the admin username and password here, but rater it should be external config.

This change doesn't really seem all that relevant to injecting a webdriver though, so complicating the review/discussions around this PR.

File: test-poc/framework/src/main/java/org/keycloak/test/framework/webdriver/WebDriverSupplier.java [hunk 4]
```diff
@@ -0,0 +1,43 @@
+package org.keycloak.test.framework.webdriver;
+
+import org.keycloak.test.framework.injection.InstanceWrapper;
+import org.keycloak.test.framework.injection.LifeCycle;
+import org.keycloak.test.framework.injection.Registry;
+import org.keycloak.test.framework.injection.Supplier;
+import org.openqa.selenium.WebDriver;
+import org.openqa.selenium.firefox.FirefoxDriver;
+
+public class WebDriverSupplier implements Supplier<WebDriver, InjectWebDriver> {
```
Comment (stianst) [inline test-poc/framework/src/main/java/org/keycloak/test/framework/webdriver/WebDriverSupplier.java:10]: Perhaps we should just call this FirefoxWebDriverSupplier? That would then go nicely together with https://github.com/orgs/keycloak/projects/42?pane=issue&itemId=63724099 as a way to select which supplier to use once we have multiple web driver suppliers implemented

File: test-poc/framework/src/main/java/org/keycloak/test/framework/server/NoAdminUserKeycloakTestServerConfig.java (hunk 2 above)
Comment (jonkoops) [reply test-poc/framework/src/main/java/org/keycloak/test/framework/server/NoAdminUserKeycloakTestServerConfig.java:3]: To set up a server without creating an administrative user, you can't test the welcome page if an administrative user already exists, it would simply redirect to the admin console.

File: test-poc/framework/src/main/java/org/keycloak/test/framework/server/DefaultKeycloakTestServerConfig.java (hunk 3 above)
Comment (jonkoops) [reply test-poc/framework/src/main/java/org/keycloak/test/framework/server/DefaultKeycloakTestServerConfig.java:7]: It's needed for https://github.com/keycloak/keycloak/pull/30378#discussion_r1647412929, it's relevant for the testing of the welcome page. If you like I can change it so `KeycloakTestServerConfig` has defaults for this, and then override these methods in `NoAdminUserKeycloakTestServerConfig` and return an empty optional.

File: test-poc/framework/src/main/java/org/keycloak/test/framework/webdriver/WebDriverSupplier.java (hunk 4 above)
Comment (jonkoops) [reply test-poc/framework/src/main/java/org/keycloak/test/framework/webdriver/WebDriverSupplier.java:10]: Yeah sure, can do. I just implemented as if it was already generic since we are missing the ability to select a specific instance at the moment.

File: test-poc/base/src/test/java/org/keycloak/test/base/WelcomePageTest.java (hunk 1 above)
Comment (jonkoops) [reply test-poc/base/src/test/java/org/keycloak/test/base/WelcomePageTest.java:13]: Yes, I'd also like to generalize this. Do you have a preference for a specific terminology? The `Test` prefix feels kinda odd to me, as it conveys no semantic meaning other than 'this is used in a test', so in this case I'd prefer `Inject` as it more accurately represents the fact this injects something into your tests. WDYT?

File: test-poc/framework/src/main/java/org/keycloak/test/framework/webdriver/WebDriverSupplier.java (hunk 4 above)
Comment (jonkoops) [reply test-poc/framework/src/main/java/org/keycloak/test/framework/webdriver/WebDriverSupplier.java]: This was implemented under af0daed939459dc7f957bac985185ff6dd1df0d5

File: test-poc/base/src/test/java/org/keycloak/test/base/WelcomePageTest.java (hunk 1 above)
Comment (stianst) [reply test-poc/base/src/test/java/org/keycloak/test/base/WelcomePageTest.java:13]: I would just call this @TestWebDriver for now, and open a separate issue to rename?

Comment (jonkoops) [reply test-poc/base/src/test/java/org/keycloak/test/base/WelcomePageTest.java:13]: Yeah sounds good, I'll log a task on the board so we can refine this later.

Comment (jonkoops) [reply test-poc/base/src/test/java/org/keycloak/test/base/WelcomePageTest.java]: Done in c689c139f66a5f8607e4aef8180ea45b4c3fde44

File: test-poc/framework/src/main/java/org/keycloak/test/framework/server/NoAdminUserKeycloakTestServerConfig.java (hunk 2 above)
Comment (stianst) [reply test-poc/framework/src/main/java/org/keycloak/test/framework/server/NoAdminUserKeycloakTestServerConfig.java:3]: Ah, got it, that makes me understand the rest - I was being a bit slow

File: test-poc/framework/src/main/java/org/keycloak/test/framework/server/KeycloakTestServerConfig.java [hunk 5]
```diff
@@ -14,4 +15,12 @@ default Set<String> features() {
         return Collections.emptySet();
     }
 
+    default Optional<String> adminUserName() {
+        return Optional.empty();
+    }
+
+    default Optional<String> adminUserPassword() {
+        return Optional.empty();
```
Comment (stianst) [inline test-poc/framework/src/main/java/org/keycloak/test/framework/server/KeycloakTestServerConfig.java:23]:
```suggestion
        return Optional.of("admin");
```

File: test-poc/framework/src/main/java/org/keycloak/test/framework/server/DefaultKeycloakTestServerConfig.java [hunk 6]
```diff
@@ -1,5 +1,17 @@
...
+    public Optional<String> adminUserName() {
+        return Optional.of("admin");
+    }
+
+    @Override
+    public Optional<String> adminUserPassword() {
+        return Optional.of("admin");
+    }
+
```
Comment (stianst) [inline test-poc/framework/src/main/java/org/keycloak/test/framework/server/DefaultKeycloakTestServerConfig.java:16]:
```suggestion
```

File: test-poc/framework/src/main/java/org/keycloak/test/framework/server/NoAdminUserKeycloakTestServerConfig.java (hunk 2 above)
Comment (stianst) [inline test-poc/framework/src/main/java/org/keycloak/test/framework/server/NoAdminUserKeycloakTestServerConfig.java:3]: Maybe move this to base module instead of framework? It's a bit of a corner case thing, so probably doesn't need to be part of the framework itself

**Review by stianst — COMMENTED** [review]
LGTM couple minor things:

* I would set the username/password in KeycloakTestServerConfig.java as I reckon the most common case is that you don't care about it, so would we good to just get the default when creating a new config
* Maybe move NoAdminUser.. to base module, as it probably doesn't need to be part of the framework

**Review by stianst — APPROVED** [review]
